diff --git a/src/gallium/state_trackers/clover/Makefile.am b/src/gallium/state_trackers/clover/Makefile.am
index ece2b38..cc9311c 100644
--- a/src/gallium/state_trackers/clover/Makefile.am
+++ b/src/gallium/state_trackers/clover/Makefile.am
@@ -37,7 +37,7 @@ libcltgsi_la_SOURCES = \
 	tgsi/compiler.cpp
 
 libclllvm_la_CXXFLAGS = \
-	-std=c++98 \
+	-std=c++11 \
 	$(VISIBILITY_CXXFLAGS) \
 	$(LLVM_CPPFLAGS) \
 	$(DEFINES) \
@@ -49,7 +49,7 @@ libclllvm_la_SOURCES = \
 	llvm/invocation.cpp
 
 libclover_la_CXXFLAGS = \
-	-std=c++0x \
+	-std=c++11 \
 	$(VISIBILITY_CXXFLAGS)
 
 libclover_la_LIBADD = \
