report:

    http://sourceware.org/ml/libffi-discuss/2011/msg00131.html

diff -ru libffi-3.0.10rc8.orig/src/closures.c libffi-3.0.10rc8/src/closures.c
--- libffi-3.0.10rc8.orig/src/closures.c	2011-05-10 08:45:31 +0200
+++ libffi-3.0.10rc8/src/closures.c	2011-05-10 09:12:33 +0200
@@ -172,7 +172,7 @@
 
 #endif /* !FFI_MMAP_EXEC_SELINUX */
 
-#elif defined (__CYGWIN__)
+#elif defined (__CYGWIN__) || defined(__INTERIX)
 
 #include <sys/mman.h>
 
