From d2cf4f57766bffeb711c740c08039ca3e3df916c Mon Sep 17 00:00:00 2001
From: Mike Lothian <mike@fireburn.co.uk>
Date: Thu, 27 Apr 2023 11:14:04 +0100
Subject: [PATCH] GCC 13 fix

---
 src/reshade/effect_token.hpp | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/reshade/effect_token.hpp b/src/reshade/effect_token.hpp
index e85e2fa..998f5f8 100644
--- a/src/reshade/effect_token.hpp
+++ b/src/reshade/effect_token.hpp
@@ -5,6 +5,7 @@
 
 #pragma once
 
+#include <cstdint>
 #include <string>
 #include <vector>
 
-- 
2.40.1

