--- a/tools/lldb/scripts/lldb_python_module.cmake
+++ b/tools/lldb/scripts/lldb_python_module.cmake
@@ -7,8 +7,8 @@
 
 SET(PYTHON_DIRECTORY python${PYTHON_VERSION_MAJOR}.${PYTHON_VERSION_MINOR})
 
-SET(lldb_module_src ${CMAKE_CURRENT_BINARY_DIR}/lib/${PYTHON_DIRECTORY})
-SET(lldb_module_dest ${CMAKE_INSTALL_PREFIX}/lib)
+SET(lldb_module_src ${CMAKE_CURRENT_BINARY_DIR}/@libdir@/${PYTHON_DIRECTORY})
+SET(lldb_module_dest ${CMAKE_INSTALL_PREFIX}/@libdir@)
 
 MESSAGE(STATUS "Installing LLDB python module from: ${lldb_module_src} to ${lldb_module_dest}")
 FILE(COPY "${lldb_module_src}" DESTINATION "${lldb_module_dest}")

--- a/tools/lldb/source/CMakeLists.txt
+++ b/tools/lldb/source/CMakeLists.txt
@@ -206,7 +206,3 @@
 endif ()
 # FIXME: implement svn/git revision and repository parsing solution on Windows. There is an SVN-only
 #        revision parsing solution in tools/clang/lib/Basic/CMakelists.txt.
-
-
-install(TARGETS liblldb
-  LIBRARY DESTINATION lib)
