From be135ce5325530853f77b576b80725cf96a854d4 Mon Sep 17 00:00:00 2001
From: Mike Lothian <mike@fireburn.co.uk>
Date: Tue, 16 Jul 2013 01:23:39 +0100
Subject: [PATCH] Add option to not install wayland-scanner

---
 configure.ac    | 8 ++++++++
 src/Makefile.am | 4 +++-
 2 files changed, 11 insertions(+), 1 deletion(-)

diff --git a/configure.ac b/configure.ac
index 536df9e..c1fd561 100644
--- a/configure.ac
+++ b/configure.ac
@@ -58,6 +58,12 @@ AC_ARG_ENABLE([scanner],
               [],
               [enable_scanner=yes])
 
+AC_ARG_ENABLE([scanner-install],
+              [AC_HELP_STRING([--disable-scanner-install],
+                              [Disable install of wayland-scanner])],
+              [],
+              [enable_scanner_install=yes])
+
 AC_ARG_ENABLE([documentation],
 	      [AC_HELP_STRING([--disable-documentation],
 		              [Disable building the documentation])],
@@ -66,6 +72,8 @@ AC_ARG_ENABLE([documentation],
 
 AM_CONDITIONAL(ENABLE_SCANNER, test "x$enable_scanner" = xyes)
 
+AM_CONDITIONAL(ENABLE_SCANNER_INSTALL, test "x$enable_scanner_install" = xyes)
+
 AC_ARG_WITH(icondir, [  --with-icondir=<dir>    Look for cursor icons here],
 		     [  ICONDIR=$withval],
 		     [  ICONDIR=${datadir}/icons])
diff --git a/src/Makefile.am b/src/Makefile.am
index 4226f63..59f4710 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -48,9 +48,11 @@ endif
 
 include $(top_srcdir)/src/scanner.mk
 
-if ENABLE_SCANNER
+if ENABLE_SCANNER_INSTALL
 bin_PROGRAMS = wayland-scanner
+endif
 
+if ENABLE_SCANNER
 wayland_scanner_SOURCES =				\
 	scanner.c
 
-- 
1.8.3.2

