--- a/configure.py	2013-02-20 19:42:06.449445817 +0000
+++ b/configure.py	2013-02-20 20:07:07.826413000 +0000
@@ -11,7 +11,11 @@
 from optparse import OptionParser
 import os
 from subprocess import *
-import sys
+import sys, platform
+
+system_libdir = 'lib32'
+if platform.machine() in ('x86_64',):
+	system_libdir = 'lib64'
 
 srcdir = os.path.dirname(sys.argv[0])
 
@@ -68,7 +72,7 @@
 llvm_link = os.path.join(llvm_bindir, 'llvm-link')
 llvm_opt = os.path.join(llvm_bindir, 'opt')
 
-default_targets = ['r600--']
+default_targets = ['r600-amd-none']
 
 targets = args
 if not targets:
@@ -168,7 +172,7 @@
   b.build(builtins_link_bc, "LLVM_LINK", objects)
   b.build(builtins_opt_bc, "OPT", builtins_link_bc)
   b.build(builtins_bc, "PREPARE_BUILTINS", builtins_opt_bc, prepare_builtins)
-  install_files_bc.append((builtins_bc, builtins_bc))
+  install_files_bc.append((builtins_bc, '%s/clc/%s' % (system_libdir, target)))
   install_deps.append(builtins_bc)
   b.default(builtins_bc)
 
