--- a/configure.py	2013-12-23 21:41:35.585238480 +0000
+++ b/configure.py	2013-12-23 21:42:25.579238656 +0000
@@ -62,9 +62,12 @@
     print "Please ensure that llvm-config is in your $PATH, or use --with-llvm-config."
     sys.exit(1)
 
+llvm_version = llvm_config(['--version']).replace("svn","")
 llvm_bindir = llvm_config(['--bindir'])
 llvm_core_libs = llvm_config(['--libs', 'core', 'bitreader', 'bitwriter']) + ' ' + \
                  llvm_config(['--ldflags'])
+if llvm_version >= '3.5':
+  llvm_core_libs = llvm_core_libs + ' ' + llvm_config(['--system-libs'])
 llvm_cxxflags = llvm_config(['--cxxflags']) + ' -fno-exceptions -fno-rtti'
 
 llvm_clang = os.path.join(llvm_bindir, 'clang')
