diff --git a/setup_vkd3d_proton.sh b/setup_vkd3d_proton.sh
index 6b0ed126..c7d264d1 100644
--- a/setup_vkd3d_proton.sh
+++ b/setup_vkd3d_proton.sh
@@ -111,17 +111,12 @@ installFile() {
   fi
 
   if [ -n "$1" ]; then
-    if [ -f "${dstfile}" ] || [ -h "${dstfile}" ]; then
       if ! [ -f "${dstfile}.old" ]; then
         mv -v "${dstfile}" "${dstfile}.old"
       else
         rm -v "${dstfile}"
       fi
       $file_cmd "${srcfile}" "${dstfile}"
-    else
-      echo "${dstfile}: File not found in wine prefix" >&2
-      return 1
-    fi
   fi
   return 0
 }
