From b0b420b8de54f748702b64c3bb0e92767bcc402c Mon Sep 17 00:00:00 2001
From: Eric Anholt <eric@anholt.net>
Date: Sat, 28 Dec 2013 01:14:41 +0000
Subject: Update for glamor in the 1.16 server.

---
diff --git a/configure.ac b/configure.ac
index 808c247..e0c6029 100644
--- a/configure.ac
+++ b/configure.ac
@@ -324,8 +324,10 @@ if test "x$GLAMOR" != "xno"; then
 	if test "x$UXA" != "xyes"; then
 		AC_MSG_ERROR([Glamor acceleration requested but UXA is not enabled])
 	fi
-	PKG_CHECK_MODULES(LIBGLAMOR, [glamor >= 0.6.0])
-	PKG_CHECK_MODULES(LIBGLAMOR_EGL, [glamor-egl])
+	if ! pkg-config --exists "xorg-server >= 1.15.99.901"; then
+		PKG_CHECK_MODULES(LIBGLAMOR, [glamor >= 0.3.1])
+		PKG_CHECK_MODULES(LIBGLAMOR_EGL, [glamor-egl])
+	fi
 	AC_DEFINE(USE_GLAMOR, 1, [Enable glamor acceleration])
 fi
 
--
cgit v0.9.0.2-2-gbebe
