diff --git a/chrome/browser/ui/autofill/payments/virtual_card_selection_dialog_controller.h b/chrome/browser/ui/autofill/payments/virtual_card_selection_dialog_controller.h
index 401717c..952bbad 100644
--- a/chrome/browser/ui/autofill/payments/virtual_card_selection_dialog_controller.h
+++ b/chrome/browser/ui/autofill/payments/virtual_card_selection_dialog_controller.h
@@ -5,6 +5,8 @@
 #ifndef CHROME_BROWSER_UI_AUTOFILL_PAYMENTS_VIRTUAL_CARD_SELECTION_DIALOG_CONTROLLER_H_
 #define CHROME_BROWSER_UI_AUTOFILL_PAYMENTS_VIRTUAL_CARD_SELECTION_DIALOG_CONTROLLER_H_
 
+#include <vector>
+
 #include "base/macros.h"
 #include "base/strings/string16.h"
 
