--- a/components/supervised_user/core/browser/family_preferences_service.cc	2023-06-08 11:05:16.482773614 +0100
+++ b/components/supervised_user/core/browser/family_preferences_service.cc	2023-06-08 11:07:26.115280294 +0100
@@ -25,10 +25,10 @@
     return head_of_household_;
   }
   const absl::optional<const Member>& GetParent() const { return parent_; }
-  const std::vector<const Member>& GetRegularMembers() const {
+  const std::vector<Member>& GetRegularMembers() const {
     return regular_members_;
   }
-  const std::vector<const Member>& GetChildren() const { return children_; }
+  const std::vector<Member>& GetChildren() const { return children_; }
 
   Family() = delete;
   explicit Family(
@@ -58,8 +58,8 @@
  private:
   absl::optional<const Member> head_of_household_;
   absl::optional<const Member> parent_;
-  std::vector<const Member> regular_members_;
-  std::vector<const Member> children_;
+  std::vector<Member> regular_members_;
+  std::vector<Member> children_;
 };
 
 }  // namespace
