diff --git a/cc/trees/layer_tree_impl.h b/cc/trees/layer_tree_impl.h
index f74d18f657c8..7cdc29e4418b 100644
--- a/cc/trees/layer_tree_impl.h
+++ b/cc/trees/layer_tree_impl.h
@@ -194,7 +194,7 @@ class CC_EXPORT LayerTreeImpl {
   // Adapts an iterator of std::unique_ptr<LayerImpl> to an iterator of
   // LayerImpl*.
   template <typename Iterator>
-  class IteratorAdapter {
+  class IteratorAdapter : public std::iterator<std::forward_iterator_tag, LayerImpl*> {
    public:
     explicit IteratorAdapter(Iterator it) : it_(it) {}
     bool operator==(IteratorAdapter o) const { return it_ == o.it_; }
