diff --git a/third_party/dav1d/BUILD.gn b/third_party/dav1d/BUILD.gn
index 788dc6cff9..151376f84d 100644
--- a/third_party/dav1d/BUILD.gn
+++ b/third_party/dav1d/BUILD.gn
@@ -25,7 +25,7 @@ if (is_win) {
 # the stack alignment, so we must use the platform's default alignment in those
 # cases; https://crbug.com/928743.
 if (current_cpu == "x86" || current_cpu == "x64") {
-  if (use_thin_lto || is_win) {
+  if (is_win) {
     needs_stack_alignment = false
     # The defaults are stack_alignment=4 for x86 and stack_alignment=16 for x64.
   } else {
@@ -42,7 +42,7 @@ if (current_cpu == "x86" || current_cpu == "x64") {
         incoming_stack_boundary = 2
       }
     } else if (current_cpu == "x64") {
-      stack_alignment = 32
+      stack_alignment = 16
 
       if (!is_clang) {
         # Values used by GCC.
