--- fontforge-20150824-old/fontforge/python.c	2016-03-20 02:03:27.842985977 +0000
+++ fontforge-20150824/fontforge/python.c	2016-03-20 02:06:29.743888063 +0000
@@ -18174,7 +18174,11 @@
     mdef->runtime.pymod_def.m_doc = mdef->docstring;
     mdef->runtime.pymod_def.m_methods = mdef->methods;
     mdef->runtime.pymod_def.m_size = -1;
+#if PY_MINOR_VERSION >= 5
+    	mdef->runtime.pymod_def.m_slots = NULL;
+#else
     mdef->runtime.pymod_def.m_reload = NULL;
+#endif
     mdef->runtime.pymod_def.m_traverse = NULL;
     mdef->runtime.pymod_def.m_clear = NULL;
     mdef->runtime.pymod_def.m_free = NULL;
