diff -ur a/src/comgr-compiler.cpp b/src/comgr-compiler.cpp
--- a/src/comgr-compiler.cpp	2022-09-08 12:17:04.624610677 +0100
+++ b/src/comgr-compiler.cpp	2022-09-08 12:15:58.229155755 +0100
@@ -245,13 +245,12 @@
                                      OPT_compress_debug_sections_EQ)) {
     if (A->getOption().getID() == OPT_compress_debug_sections) {
       // TODO: be more clever about the compression type auto-detection
-      Opts.CompressDebugSections = llvm::DebugCompressionType::GNU;
+      Opts.CompressDebugSections = llvm::DebugCompressionType::None;
     } else {
       Opts.CompressDebugSections =
           llvm::StringSwitch<llvm::DebugCompressionType>(A->getValue())
               .Case("none", llvm::DebugCompressionType::None)
               .Case("zlib", llvm::DebugCompressionType::Z)
-              .Case("zlib-gnu", llvm::DebugCompressionType::GNU)
               .Default(llvm::DebugCompressionType::None);
     }
   }
@@ -454,7 +453,7 @@
     std::unique_ptr<MCCodeEmitter> MCE;
     std::unique_ptr<MCAsmBackend> MAB;
     if (Opts.ShowEncoding) {
-      MCE.reset(TheTarget->createMCCodeEmitter(*MCII, *MRI, Ctx));
+      MCE.reset(TheTarget->createMCCodeEmitter(*MCII, Ctx));
       MCTargetOptions Options;
       MAB.reset(TheTarget->createMCAsmBackend(*STI, *MRI, Options));
     }
@@ -473,7 +472,7 @@
       Out = BOS.get();
     }
 
-    MCCodeEmitter *CE = TheTarget->createMCCodeEmitter(*MCII, *MRI, Ctx);
+    MCCodeEmitter *CE = TheTarget->createMCCodeEmitter(*MCII, Ctx);
     MCTargetOptions Options;
     MCAsmBackend *MAB = TheTarget->createMCAsmBackend(*STI, *MRI, Options);
     Triple T(Opts.Triple);
